File: LeetCode/Medium/841 - Keys and Rooms/841 - Keys and Rooms.js
/**
 * @param {number[][]} rooms
 * @return {boolean}
 */
 var canVisitAllRooms = function(rooms) {
    const n = rooms.length;
    const vis = Array(n);
    
    const solve = (room) => {
        if (vis[room]) return;
        vis[room] = true;
        
        for (let key of rooms[room]) solve(key);
    };
    solve(0);
    
    for (let v of vis) {
        if (!v) return false;
    }
    return true;
};

File: LeetCode/Medium/841 - Keys and Rooms/841 - Keys and Rooms.cpp
class Solution {
private:
    int n;
    vector<bool> vis;
public:
    bool canVisitAllRooms(vector<vector<int>>& rooms) {
        n = rooms.size();
        vis = vector<bool>(n);
        
        solve(rooms, 0);
        
        for (auto room : vis) {
            if (!room) return false;
        }
        return true;
    }
    
    void solve(vector<vector<int>> &rooms, int room) {
        if (vis[room]) return;
        vis[room] = true;
        
        for (auto key : rooms[room]) solve(rooms, key);
    }
};